modules/message-lane/src/lib.rs

```diff
@@ -90,19 +159,237 @@ decl_module! {
+			let messages_dispatch_weight: Weight = messages.iter().map(|m| dispatcher.dispatch_weight(m)).sum();
+
+			// the submitter pays for dispatch in advance, so the whole `messages_dispatch_weight` should
+			// be covered. But it has already paid regular fee for at least `max_dispatch_weight`, so we
+			// only need to withdraw `messages_dispatch_weight - max_dispatch_weight`.
+			let messages_dispatch_extra_weight = messages_dispatch_weight.saturating_sub(max_dispatch_weight);
+			T::MessageDispatchPayment::pay_dispatch_fee(&relayer, messages_dispatch_extra_weight).map_err(|err| {
+				frame_support::debug::trace!(
+					target: "runtime",
+					"Rejecting {} messages because relayer {:?} is unable to pay extra fee for weight {}: {:?}",
+					messages.len(),
+					relayer,
+					messages_dispatch_weight,
+					err,
+				);
+
+				Error::<T, I>::FailedToWithdrawDispatchFee
+			})?;
+
+			// now let's process as many messages, as possible
+			let total_messages = messages.len();
+			let mut invalid_messages = 0;
+			let mut processed_messages = 0;
+			let mut queued_messages = 0;
+			for message in messages {
+				let mut lane = inbound_lane::<T, I>(message.key.lane_id);
+				match lane.receive_message(message.key.nonce, message.data, &mut dispatcher) {
+					ReceiveMessageResult::Invalid => invalid_messages += 1,
+					ReceiveMessageResult::Processed => processed_messages += 1,
+					ReceiveMessageResult::Queued => queued_messages += 1,
+				}
+			}
+
+			let weight_left = dispatcher.weight_left();
+			frame_support::debug::trace!(
+				target: "runtime",
+				"Received messages: total={}, invalid={}, processed={}, queued={}. Weight spent: {}, left: {}",
+				total_messages,
+				invalid_messages,
+				processed_messages,
+				queued_messages,
+				max_dispatch_weight.saturating_sub(weight_left),
+				weight_left,
+			);
+
+			// ok - we have processed or queued all bundled messages. But what if relayer acts aggressivly
+			// (i.e. `max_dispatch_weight` is larger than `messages_dispatch_weight`), or if some messages
+			// have been queued? Then we may have some capacity left from `max_dispatch_weight`.
+			// => let's not waste it and try to process queued messages
+			let mut queued_dispatcher = T::QueuedMessageDispatch::with_allowed_weight(weight_left);
+			let (queued_messages_dispatched, weight_spent_on_queued_dispatch) = queued_dispatcher.dispatch();
+			if queued_messages_dispatched != 0 {
+				frame_support::debug::trace!(
+					target: "runtime",
+					"Spent {} weight on {} queued messages from receive-messages",
+					weight_spent_on_queued_dispatch,
+					queued_messages_dispatched,
+				);
+			}
+
+			Ok(())
+		}
+
+		/// Receive messages receiving proof from bridged chain.
+		#[weight = 0] // TODO: update me (https://github.com/paritytech/parity-bridges-common/issues/78)
+		pub fn receive_message_receiving_proof(origin, proof: MessagesReceivingProofOf<T, I>) -> DispatchResult {
```

Comment:
```suggestion
		pub fn receive_message_delivery_proof(origin, proof: MessagesDeliveredProofOf<T, I>) -> DispatchResult {
```
Why is it `MessagesReceiving` and not `MessagesDelivered`?